Src/SwqlStudio/CrudTab.cs

```diff
@@ -0,0 +1,160 @@
+            {
+                var y = 0;
+
+                // for update/delete, accept uri
+                if (_operation == CrudOperation.Update || _operation == CrudOperation.Delete)
+                {
+                    _uri = CreateLabelAndTextbox("SwisUri", ref y);
+                    y += ControlsMargin * 3;
+                }
+
+                // for delete, uri is enough
+                if (_operation != CrudOperation.Delete)
+                {
+                    foreach (var property in _entity.Properties)
+                    {
+                        var textBox = CreateLabelAndTextbox(property.Name, ref y);
+
+                        _textboxes[property] = textBox;
+                    }
+                }
+
+                var commit = new Button();
+                container.Controls.Add(commit);
+
+                commit.Text = _operation.ToString();
+                commit.Top = y;
+                commit.AutoSize = true;
+                commit.Click += OnCommit;
+            }
+            finally
+            {
+                container.ResumeLayout();
+            }
+        }
+
+        private void OnCommit(object sender, EventArgs e)
+        {
+            var propertyBag = CreatePropertyBag();
+
+            try
+            {
+                var message = "";
+                switch (_operation)
+                {
+                    case CrudOperation.Create:
+                        var rv = ConnectionInfo.Proxy.Create(_entity.FullName, propertyBag);
+                        message = $"Created: {rv}";
+                        break;
+                    case CrudOperation.Update:
+                        ConnectionInfo.Proxy.Update(_uri.Text, propertyBag);
+                        message = "OK";
+                        break;
+                    case CrudOperation.Delete:
+                        ConnectionInfo.Proxy.Delete(_uri.Text);
+                        message = "OK";
+                        break;
+                    default:
+                        throw new ArgumentOutOfRangeException();
+                }
+
+                MessageBox.Show(message, "SwqlStudio", MessageBoxButtons.OK, MessageBoxIcon.Information);
+                OnCloseItself();
+            }
+            catch (Exception ex)
+            {
+                MessageBox.Show($"Failed to {_operation}: {ex.Message}",
+                    "SwqlStudio",
+                    MessageBoxButtons.OK,
+                    MessageBoxIcon.Error);
+            }
+        }
+
+        private PropertyBag CreatePropertyBag()
+        {
+            var rv = new PropertyBag();
+            foreach (var property in _textboxes)
+            {
+                if (!string.IsNullOrEmpty(property.Value.Text))
+                {
+                    rv[property.Key.Name] = property.Value.Text;
+                }
+            }
+
+            return rv;
+        }
+
+
+        private TextBox CreateLabelAndTextbox(string name, ref int y)
```

Comment:
When you want to stack a bunch of dynamically-created winforms controls, the `Dock = DockStyle.Top` option works well too.

Reply:
Thanks for the tip. I tried refactoring, but it didn't work as I expected, I'll keep this ugly solution.